### ISSUE TITLE
Nvmf spdk target is not working with spdk bdev via SoftRoCE

### ISSUE THREAD

**Issue by vinayakhegdev** [spdk/spdk#191]
I am trying to use spdk malloc lun over nvmf subsystem in virtual mode.
Used Soft ROCE for this.
It works fine with nvme-cli initiator.
But when I try to use bdevperf test on this, it is not working, (worked only once see the below log).

Spdk version used:
17.07.
~~~
root@vinayak-ThinkPad-L440:~/work/code/spdk/test/lib/bdev# uname -a
Linux vinayak-ThinkPad-L440 4.10.0-33-generic #37~16.04.1-Ubuntu SMP Fri Aug 11 14:07:24 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
root@vinayak-ThinkPad-L440:~/work/code/spdk/test/lib/bdev# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.2 LTS
Release:	16.04
Codename:	xenial
root@vinayak-ThinkPad-L440:~/work/code/spdk/test/lib/bdev# 
~~~
configuration nvmf server side:
~~~
[Subsystem1]
  NQN nqn.2016-06.io.spdk:cnode1
  Core 0
  Mode Virtual
  Listen RDMA 172.16.10.90:4420
  #Host nqn.2016-06.io.spdk:init
  SN SPDK00000000000001
  Namespace Malloc0
~~~
Client side:
~~~
[Nvme]
  TransportID "trtype:RDMA adrfam:IPv4 subnqn:nqn.2016-06.io.spdk:cnode1 traddr:172.16.10.90 trsvcid:4420" Nvme1
~~~

~~~
Occupied cpu socket mask is 0x1
reactor.c: 314:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
Running I/O for 1 seconds...
 Logical core: 0
 Nvme1n1             :    3185.00 IO/s       1.56 MB/s
 =====================================================
 Total               :    3185.00 IO/s       1.56 MB/s
done.
root@vin-VirtualBox:~/code/spdk/test/lib/bdev# 
root@vin-VirtualBox:~/code/spdk/test/lib/bdev# ./bdevperf/bdevperf -c kpc-bdev.conf -q 1 -s 512 -w verify -t 1
Starting DPDK 17.05.0 initialization...
[ DPDK EAL parameters: bdevtest -c 0x1 --file-prefix=spdk_pid2617 ]
EAL: Detected 2 lcore(s)
EAL: Probing VFIO support...
Occupied cpu socket mask is 0x1
reactor.c: 314:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
Running I/O for 1 seconds...
nvme_rdma.c: 184:nvme_rdma_get_event: *ERROR*: Received event 8 from CM event channel, but expected event 9
nvme_rdma.c: 510:nvme_rdma_connect: *ERROR*: RDMA connect error
nvme_rdma.c: 746:nvme_rdma_qpair_connect: *ERROR*: Unable to connect the rqpair
nvme_ctrlr.c: 233:spdk_nvme_ctrlr_alloc_io_qpair: *ERROR*: transport->ctrlr_create_io_qpair() failed
Segmentation fault (core dumped)
~~~

Backtrace.
~~~
root@vin-VirtualBox:~/code/spdk/test/lib/bdev# 
Current thread is 1 (Thread 0x7fa7fc51b8c0 (LWP 12788))]
(gdb) bt
#0  _bdev_nvme_submit_request (bdev_io=0x7fa7d9a84d00, ch=0x0) at bdev_nvme.c:361
#1  bdev_nvme_submit_request (ch=0x0, bdev_io=0x7fa7d9a84d00) at bdev_nvme.c:422
#2  0x00000000004195a0 in __submit_request (bdev=0x1fa4a40, bdev_io=0x7fa7d9a84d00)
    at bdev.c:571
#3  spdk_bdev_io_submit (bdev_io=0x7fa7d9a84d00) at bdev.c:580
#4  spdk_bdev_writev (desc=desc@entry=0x1fa4c30, ch=ch@entry=0x1fa4c70, iov=0x7fa7d98c0b00, 
    iovcnt=iovcnt@entry=1, offset=0, len=len@entry=4096, 
    cb=0x4056a0 <bdevperf_verify_write_complete>, cb_arg=0x7fa7d98c0b00) at bdev.c:923
#5  0x0000000000405e51 in bdevperf_submit_single (target=target@entry=0x1f169a0)
    at bdevperf.c:376
#6  0x00000000004060fb in bdevperf_submit_io (queue_depth=0, target=0x1f169a0)
    at bdevperf.c:393
#7  bdevperf_submit_on_core (arg1=<optimized out>, arg2=<optimized out>) at bdevperf.c:469
#8  0x000000000041bb92 in _spdk_event_queue_run_batch (reactor=0x660280 <g_reactors>, 
    reactor=0x660280 <g_reactors>) at reactor.c:195
#9  _spdk_reactor_run (arg=arg@entry=0x660280 <g_reactors>) at reactor.c:324
#10 0x000000000041c09d in spdk_reactors_start () at reactor.c:523
#11 0x000000000041b4a2 in spdk_app_start (opts=opts@entry=0x7ffca9de6b50, 
    start_fn=start_fn@entry=0x4057f0 <bdevperf_run>, arg1=arg1@entry=0x0, 
    arg2=arg2@entry=0x0) at app.c:377
#12 0x0000000000404c62 in main (argc=11, argv=0x7ffca9de6cc8) at bdevperf.c:737
(gdb) 
~~~

**Comment by optimistyzy** [comment]
@vinayakhegdev  Can you check with the latest SPDK code and see whether this happens again?

**Comment by vinayakhegdev** [comment]
Checked with the latest code.
It works sometime only, not always.
This is the client log when it failed.

root@vin-VirtualBox:~/code/spdk-latest/spdk/test/lib/bdev# ./bdevperf/bdevperf -c ~/code/spdk/test/lib/bdev/kpc-bdev.conf -q 1 -s 4096 -w write -t 1
Starting DPDK 17.05.0 initialization...
[ DPDK EAL parameters: bdevtest -c 0x1 -m 1024 --file-prefix=spdk_pid2661 ]
EAL: Detected 2 lcore(s)
EAL: Probing VFIO support...
Total cores available: 1
Occupied cpu socket mask is 0x1
reactor.c: 362:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
Running I/O for 1 seconds...
 Logical core: 0
 Nvme1n1             :    3393.00 IO/s      13.25 MB/s
 =====================================================
 Total               :    3393.00 IO/s      13.25 MB/s


nvme_rdma.c:1487:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x170d0e0, Response Index 24213824 (12): transport retry counter exceeded
nvme_rdma.c:1487:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x170d0e0, Response Index 21 (5): Work Request Flushed Error


=============================================================

Server log:

rdma.c:1707:spdk_nvmf_rdma_qpair_poll: *ERROR*: CQ error on CQ 0x7f6aa0010180, Request 0x140095927627152 (12): transport retry counter exceeded
rdma.c:1707:spdk_nvmf_rdma_qpair_poll: *ERROR*: CQ error on CQ 0x7f6aa0010180, Request 0x140095928027928 (5): Work Request Flushed Error
rdma.c:1707:spdk_nvmf_rdma_qpair_poll: *ERROR*: CQ error on CQ 0x7f6aa0010180, Request 0x140095928028016 (5): Work Request Flushed Error
rdma.c:1707:spdk_nvmf_rdma_qpair_poll: *ERROR*: CQ error on CQ 0x7f6aa0010180, Request 0x140095928028104 (5): Work Request Flushed Error
rdma.c:1707:spdk_nvmf_rdma_qpair_poll: *ERROR*: CQ error on CQ 0x7f6aa0010180, Request 0x140095928028192 (5): Work Request Flushed Error

**Comment by optimistyzy** [comment]
@vinayakhegdev  will look at it.

**Comment by vinayakhegdev** [comment]
Tried disabling it, but get the same error.

root@vin-VirtualBox:~/code/spdk-latest/spdk/test/lib/bdev# ./bdevperf/bdevperf -c kpc-bdev.conf -q 1 -s 4096 -w write -t 1
Starting DPDK 17.05.0 initialization...
[ DPDK EAL parameters: bdevtest -c 0x1 -m 1024 --file-prefix=spdk_pid2682 ]
EAL: Detected 2 lcore(s)
EAL: Probing VFIO support...
Total cores available: 1
Occupied cpu socket mask is 0x1
reactor.c: 362:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
nvme_rdma.c:1487:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0xdc2870, Response Index 19257408 (12): transport retry counter exceeded
nvme_rdma.c:1487:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0xdc2870, Response Index 11 (5): Work Request Flushed Error

**Comment by optimistyzy** [comment]
@vinayakhegdev  I also reproduced this issue in my platform, and would have  a detailed analysis on whether it is the software roce issue or SPDK issue.

**Comment by vinayakhegdev** [comment]
Anything found on this?

**Comment by optimistyzy** [comment]
not finding any thing wrong in spdk yet,perhaps the software roce issue

**Comment by optimistyzy** [comment]
btw, see someone post related issue on software roce homepage in github. Need to identify root cause.

**Comment by vinayakhegdev** [comment]
Ok.
So Can you please tell me which hardware I can use which is very well tested with SPDK.
Which RNIC adapter?
Which Nvme disk?

Does it work with Intel 3d Xpoint disks, if so can you specify which one?
If possible can you tell me the latency numbers you guys got with spdk?
That would be helpful for me to further invest my time on spdk and its code.

**Comment by vinayakhegdev** [comment]
Can you please answer.

**Comment by optimistyzy** [comment]
@vinayakhegdev  SPDK NVMf target can work with any Intel 3D Xpoint SSDs (e.g., P4800X).  For the latency number, it depends on the configuration and platform.  For some materials, you can view it in the following link (http://www.spdk.io/news/2017/05/03/summit_presentations/)

**Comment by sungjoonahn** [comment]
Hi @vinayakhegdev, 

I'm trying to use SPDK+NVMe-oF_SoftRoCE. The only setup that worked for me is SPDK nvmf target and nvme-cli initiator on the same machine over a loopback network connection. I tried multiple target/initiator combinations over two physical machines but hasn't been successful in creating NVMe-oF connection over SoftRoCE. 

Do you have a NVMe-oF / SoftRoCE setup that works over two physical machines? If so, could you share the information about OS, NIC, SPDK version, etc.? 

Thanks,
Sungjoon

**Comment by vinayakhegdev** [comment]
I didnt try with physical machine, I tried with 2 virtual machine.
It is not working properly, every now and then it crashes.
OS: Ubuntu 16.04
Spdk : 17.05

**Comment by sungjoonahn** [comment]
@vinayakhegdev Thanks for your reply.

**Comment by optimistyzy** [comment]
Some updates: @vinayakhegdev @sungjoonahn  Hoping the following can help you to do some basic tests on soft roce.

I used the following spdk version (commit e35816b4be222a6cf57204fe89bacac7475826fc)  and Linux kernel 4.14-rc2 (commit 9cd6681cb1169e815c41af0265165dd1b872f228) and with the attached patch file (actually it is not a patch, just to make the basic NVMe-oF target/host function work with softroce, I do believe there are defects in rxe module in kernel, and call for help to solve the defects).   And I currently can run basic read/write test.  For read, I can use any I/O size. For write, I can use limited I/O size (e.g., 512, 1024). Currently, I did not use verify mode, so the current test for SPDK NVME-oF target/host test is very basic. 

Generally, I currently think that soft roce is still not robust enough to run complicated NVMe-oF target/host. 

BTW: I used two physical machines, and they are directly connected. 
[Uploading patch.txt…]

**Comment by sudhesam** [comment]
Hello,

Below is the details of my setup.

I have 2 Ubuntu VM's running as initiator (SudheUbu1) and target (SudheUbu2).  I am using SoftRoCE based initiator -> target data transfer.  

I have created an 'emulated' NVMe device on the target and exported the device as SPDK device.  Below is the configuration file on the target:

root@SudheUbu2:/home/ssgroot/GIT_REPOS/spdk/etc/spdk# lspci | grep mem
0b:00.0 Non-Volatile memory controller: VMware Device 07f0

root@SudheUbu2:/home/ssgroot/GIT_REPOS/spdk/etc/spdk# grep Transport nvmf.conf
  TransportId "trtype:PCIe traddr:0000:0b:00.0" Nvme1

Below is my subsystem configuration setting on the target :

# Namespaces backed by physical NVMe devices
[Subsystem1]
  NQN nqn.2016-06.io.spdk:cnode1
  Listen RDMA 192.168.123.2:4420
  AllowAnyHost Yes
  Host nqn.2016-06.io.spdk:init
  SN SPDK00000000000001
#  Namespace AIO0
#  AllowAnyHost No
#  Host nqn.2016-06.io.spdk:init
#  SN SPDK00000000000001
#  Namespace Nvme0n1 1
  Namespace Nvme1n1 1


On the initiator side, I am able to see the device as SPDK target.  Below is the output of 'nvme list' command :

root@SudheUbu1:/tmp/ssam# nvme list
Node             SN                   Model                                    Namespace Usage                      Format           FW Rev
---------------- -------------------- ---------------------------------------- --------- -------------------------- ---------------- --------
/dev/nvme0n1     VMWare NVME-0000     Virtual disk                             1          10.74  GB /  10.74  GB    512   B +  0 B   1.0
/dev/nvme1n1     SPDK00000000000001   SPDK bdev Controller                     1          10.74  GB /  10.74  GB    512   B +  0 B   18.01
root@SudheUbu1:/tmp/ssam#

Now, when I run 'identify' application, I get the following error :

root@SudheUbu1:/tmp/ssam# ./identify -r 'trtype:RDMA adrfam:IPv4 traddr:192.168.123.2 trsvcid:4420 subnqn:nqn.2016-06.io.spdk:cnode1'
Starting DPDK 17.02.0 initialization...
[ DPDK EAL parameters: identify -c 0x1 -n 1 -m 64 --no-pci --file-prefix=spdk_pid4845 ]
EAL: Detected 2 lcore(s)
EAL: No free hugepages reported in hugepages-1048576kB
nvme.c: 417:spdk_nvme_probe: *ERROR*: NVMe trtype 1 not available
spdk_nvme_probe() failed

But, when I run the 'perf' application, I get the following :

root@SudheUbu1:/tmp/ssam# ./perf -q 1 -s 4096 -w read -M 100 -t 60 -r 'trtype:RDMA adrfam:IPv4 traddr:192.168.123.2 trsvcid:4420 subnqn:nqn.2016-06.io.spdk:cnode1' -d 2048
Ignoring -M option... Please use -M option only when using rw or randrw.
Starting DPDK 17.02.0 initialization...
[ DPDK EAL parameters: perf -c 0x1 -m 2048 --no-pci --file-prefix=spdk_pid4836 ]
EAL: Detected 2 lcore(s)
EAL: No free hugepages reported in hugepages-1048576kB
Initializing NVMe Controllers
Attaching to NVMe over Fabrics controller at 192.168.123.2:4420: nqn.2016-06.io.spdk:cnode1
Attached to NVMe over Fabrics controller at 192.168.123.2:4420: nqn.2016-06.io.spdk:cnode1
Associating SPDK bdev Controller (SPDK00000000000001  ) with lcore 0
Initialization complete. Launching workers.
Starting thread on core 0
nvme_rdma.c:1448:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x55fc1f04cbc0, Response Index 94541342743896 (12): transport retry counter exceeded
nvme_rdma.c:1448:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x55fc1f04cbc0, Response Index 1 (5): Work Request Flushed Error

Any help appreciated.

**Comment by benlwalker** [comment]
Soft RoCE with NVMe-oF in loopback was broken for awhile in the kernel itself. I believe it was fixed in 4.14. I regularly use 4.14 with soft RoCE and everything works great. Can you confirm that using a very recent kernel fixes the problem?

**Comment by ssampath** [comment]
Thank you Ben for your reply.

Couple of things:

1.  I am not using loopback since I have 2 separate VM's running Soft RoCE.
2.  Our environment is currently using 4.9 kernel version.  (Hands currently tied up from moving to a 
     newer kernel).

Having said that, any help to identify the kernel patches that went it between 4.9 -> 4.14 related to     
Soft RoCE/RDMA/IB, I can patch those and try to see if it works.

Thanks again for all the help and support.

**Comment by dverkamp-intel** [comment]
I think NVMe-oF with SoftRoCE should be working in the latest SPDK + Linux environment (we test this in our automated build pool on the VM named fedora-06), so I'm going to close this issue; if new problems come up with SoftRoCE, please reopen or create a new issue.
